File: services/vios/src/framework/notification/redis/redis_subscriber.cpp
/*
 * SPDX-FileCopyrightText: Copyright (c) 2023-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <dlfcn.h>
#include <cstdlib>
#include <jsoncpp/json/json.h>
#include "redis_subscriber.h"
#include "config.h"
#include "logger.h"
#include "utils.h"
#include "ds_proto_parser.h"

using namespace std;

static void* openLibrary(const char* libName)
{
    std::string lib_path;
    void* handle = nullptr;

#if defined(AARCH64_PLATFORM)
    lib_path = std::string(ABSOLUTE_PREBUILT_LIBRARY_PATH_ARCH64) + libName;
    handle = dlopen(lib_path.c_str(), RTLD_LAZY);
    if (!handle)
    {
        lib_path = std::string(RELATIVE_PREBUILT_LIBRARY_PATH_ARCH64) + libName;
        handle = dlopen(lib_path.c_str(), RTLD_LAZY);
    }
#else
    lib_path = std::string(ABSOLUTE_PREBUILT_LIBRARY_PATH_X86_64) + libName;
    handle = dlopen(lib_path.c_str(), RTLD_LAZY);
#endif

    return handle;
}

static void subscribe_cb(NvDsMsgApiErrorType flag, void *msg, int len, char *topic, void *user_ptr)
{
    if(flag == NVDS_MSGAPI_ERR)
    {
        LOG(error) << "Error in consuming message from redis broker" << endl;
    }
    else
    {
        RedisSubscriber* subscriber = (RedisSubscriber*) user_ptr;
        if (subscriber)
        {
            subscriber->deliverMessage(msg, len);
        }
    }
}

RedisSubscriber* RedisSubscriber::_instance = nullptr;

RedisSubscriber* RedisSubscriber::getInstance()
{
    if (_instance == nullptr)
    {
        _instance = new RedisSubscriber();
    }
    return _instance;
}

void RedisSubscriber::deleteInstance()
{
    if (_instance != nullptr)
    {
        delete _instance;
        _instance = nullptr;
    }
}

RedisSubscriber::RedisSubscriber()
        : nvds_msgapi_connect(nullptr)
        , nvds_msgapi_subscribe(nullptr)
        , nvds_msgapi_disconnect(nullptr)
        , m_connHandle(nullptr)
        , m_handleRedis(nullptr)
        , m_handleRedisProto(nullptr)
        , m_subscribeTopic("")
        , m_error(false)
{
    // Temporary solution to load libnvds_logger in memory
    m_handleRedis = openLibrary("libnvds_logger.so");
    if (!m_handleRedis)
    {
        LOG(error) << "Cannot open nvds_logger library: " << dlerror() << endl;
        goto error;
    }

    m_handleRedisProto = openLibrary("libnvds_redis_proto.so");
    if (!m_handleRedisProto)
    {
        LOG(error) << "Cannot open nvds_redis library: " << dlerror() << endl;
        goto error;
    }

    dlerror();

    nvds_msgapi_connect = (msgapi_connect_ptr) dlsym(m_handleRedisProto, "nvds_msgapi_connect");
    nvds_msgapi_subscribe = (msgapi_subscribe_ptr) dlsym(m_handleRedisProto, "nvds_msgapi_subscribe");
    nvds_msgapi_disconnect = (msgapi_disconnect_ptr) dlsym(m_handleRedisProto, "nvds_msgapi_disconnect");

    if (nvds_msgapi_connect == nullptr || nvds_msgapi_subscribe == nullptr || nvds_msgapi_disconnect == nullptr)
    {
        goto error;
    }

    /* redis subscriber initialization */
    redisInit();

    /* Initialize DS Proto Parser */
    DsProtoParser::getInstance();
    return;

error:
    const char *dlsym_error = dlerror();
    if (dlsym_error)
    {
        LOG(error) << "Cannot load symbol 'nvds_msgapi_': " << dlsym_error << endl;
        if (m_handleRedis)
        {
            dlclose(m_handleRedis);
        }
        if (m_handleRedisProto)
        {
            dlclose(m_handleRedisProto);
        }
        m_error = true;
    }
}

RedisSubscriber::~RedisSubscriber()
{
    nvds_msgapi_disconnect(m_connHandle);
    if (m_handleRedis)
    {
        dlclose(m_handleRedis);
    }
    if (m_handleRedisProto)
    {
        dlclose(m_handleRedisProto);
    }
    m_listeners.clear();
}

void RedisSubscriber::redisInit()
{
    m_subscribeTopic = GET_CONFIG().message_broker_topic_consumer;
    const char *topic[] = {m_subscribeTopic.c_str()};

    m_redisEndpoint = getRedisServerEndpoint();
    // Connect to Redis server
    m_connHandle = nvds_msgapi_connect((char*)m_redisEndpoint.c_str(),
                                        nullptr, (char*)"");
    if (!m_connHandle)
    {
        LOG(error) << "Redis Subscriber Connect failed. Exiting" << endl;
        goto error;
    }
    LOG(info) << "Redis Subscriber connect success." << endl;

    //Subscribe to topic
    if(nvds_msgapi_subscribe(m_connHandle, (char **)topic, 1, subscribe_cb, (void*)this) != NVDS_MSGAPI_OK)
    {
        LOG(error) << "Redis subscription to topic[s] failed. Exiting" << endl;
        goto error;
    }
    return;
error:
    m_error = true;
}

void RedisSubscriber::registerMessageListener(nv_vms::INotificationListener* listener)
{
    LOG(info) << "Registering message listener" << endl;
    std::lock_guard<std::mutex> lock(m_listenerMutex);
    m_listeners.insert(listener);
}

void RedisSubscriber::deregisterMessageListener(nv_vms::INotificationListener* listener)
{
    LOG(info) << "Deregistering message listener" << endl;
    std::lock_guard<std::mutex> lock(m_listenerMutex);
    m_listeners.erase(listener);
}

void RedisSubscriber::deliverMessage(void *msg, int len)
{
    if (!m_listeners.size())
    {
        return;
    }

    int64_t frameTimeMs = 0;
    Json::Value payload = DsProtoParser::getInstance()->parseMessage(msg, len, frameTimeMs);
    if (payload == Json::nullValue)
    {
        static std::atomic<uint64_t> logError{0};
        if (logError < 10)
        {
            LOG(error) << "Unable to parse " << logError << " messages" << endl;
            logError++;
        }
        return;
    }

    std::lock_guard<std::mutex> lock(m_listenerMutex);
    for (auto listener: m_listeners)
    {
        if (listener == nullptr)
        {
            continue;
        }
        if (payload.isArray())
        {
            for (Json::ArrayIndex i = 0; i < payload.size(); i++)
            {
                listener->onMessage(payload[i]);
            }
        }
        else
        {
            listener->onMessage(payload);
        }
    }
}

File: services/sdrc/lib/messaging/redis_subscriber.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import logging
import signal
import time
from typing import Callable, Optional, Sequence

import redis


class GracefulExit:
    def __init__(self, enable_signals: bool = True):
        self.should_exit = False
        if enable_signals:
            try:
                signal.signal(signal.SIGINT, self._handle)
                signal.signal(signal.SIGTERM, self._handle)
            except ValueError:
                # Signals can only be set in the main thread; ignore in workers
                pass

    def _handle(self, *_):
        self.should_exit = True


class RedisSubscriber:
    def __init__(
        self,
        redis_client: redis.StrictRedis,
        logger: Optional[logging.Logger] = None,
        ignore_subscribe_messages: bool = True,
        enable_signals: bool = True,
    ) -> None:
        self._redis = redis_client
        self._pubsub = self._redis.pubsub(
            ignore_subscribe_messages=ignore_subscribe_messages
        )
        self._logger = logger or logging.getLogger(__name__)
        self._stopper = GracefulExit(enable_signals=enable_signals)

    def subscribe(self, channels: Sequence[str]) -> None:
        if channels:
            self._logger.info("Subscribing to channels: %s", ", ".join(channels))
            self._pubsub.subscribe(*channels)

    def psubscribe(self, patterns: Sequence[str]) -> None:
        if patterns:
            self._logger.info("Pattern-subscribing: %s", ", ".join(patterns))
            self._pubsub.psubscribe(*patterns)

    def listen(
        self,
        handler: Callable[[str, str], None],
        poll_sleep_seconds: float = 0.05,
    ) -> None:
        """Start consuming pub/sub messages.

        handler: callable taking (channel, data_str)
        """
        self._logger.info("Starting Redis pub/sub listener")
        while not self._stopper.should_exit:
            try:
                message = self._pubsub.get_message(timeout=1.0)
                if not message:
                    time.sleep(poll_sleep_seconds)
                    continue

                mtype = message.get("type")
                if mtype not in ("message", "pmessage"):
                    continue

                channel = message.get("channel")
                data = message.get("data")
                if not isinstance(data, str):
                    try:
                        data = json.dumps(data)
                    except Exception:
                        data = str(data)

                handler(channel, data)
            except KeyboardInterrupt:
                break
            except Exception:
                self._logger.exception("Error while consuming pub/sub message")
                time.sleep(0.25)

    def close(self) -> None:
        try:
            self._pubsub.close()
        except Exception:
            pass

    def stop(self) -> None:
        self._stopper.should_exit = True
        self.close()

